src/System.EventLog/src/System/Diagnostics/SharedUtils.cs

```diff
@@ -0,0 +1,335 @@
+            {
+                if (environment == UnknownEnvironment)
+                {
+                    lock (InternalSyncObject)
+                    {
+                        if (environment == UnknownEnvironment)
+                        {
+                            if (Environment.OSVersion.Platform == PlatformID.Win32NT)
+                            {
+                                if (Environment.OSVersion.Version.Major >= 5)
+                                    environment = W2kEnvironment;
+                                else
+                                    environment = NtEnvironment;
+                            }
+                            else
+                                environment = NonNtEnvironment;
+                        }
+                    }
+                }
+
+                return environment;
+            }
+        }
+
+        internal static void CheckEnvironment()
+        {
+            if (CurrentEnvironment == NonNtEnvironment)
+                throw new PlatformNotSupportedException(SR.WinNTRequired);
+        }
+
+        internal static void CheckNtEnvironment()
+        {
+            if (CurrentEnvironment == NtEnvironment)
+                throw new PlatformNotSupportedException(SR.Win2000Required);
+        }
+
+        internal static void EnterMutex(string name, ref Mutex mutex)
+        {
+            string mutexName = null;
+            if (CurrentEnvironment == W2kEnvironment)
+                mutexName = "Global\\" + name;
+            else
+                mutexName = name;
+
+            EnterMutexWithoutGlobal(mutexName, ref mutex);
+        }
+
+        internal static void EnterMutexWithoutGlobal(string mutexName, ref Mutex mutex)
+        {
+            bool createdNew;
+            MutexSecurity sec = new MutexSecurity();
+            SecurityIdentifier everyoneSid = new SecurityIdentifier(WellKnownSidType.AuthenticatedUserSid, null);
+            //sec.AddAccessRule(new MutexAccessRule(everyoneSid, MutexRights.Synchronize | MutexRights.Modify, AccessControlType.Allow));
+            //Mutex tmpMutex = new Mutex(false, mutexName, out createdNew, sec);
+            Mutex tmpMutex = new Mutex(false, mutexName, out createdNew);
+            SafeWaitForMutex(tmpMutex, ref mutex);
+        }
+
+        private static bool SafeWaitForMutex(Mutex mutexIn, ref Mutex mutexOut)
+        {
+            Debug.Assert(mutexOut == null, "You must pass in a null ref Mutex");
+            while (true)
+            {
+                if (!SafeWaitForMutexOnce(mutexIn, ref mutexOut))
+                    return false;
+                if (mutexOut != null)
+                    return true;
+                Thread.Sleep(0);
+            }
+        }
+
+        [MethodImplAttribute(MethodImplOptions.NoInlining)]
+        private static bool SafeWaitForMutexOnce(Mutex mutexIn, ref Mutex mutexOut)
+        {
+            bool ret;
+            RuntimeHelpers.PrepareConstrainedRegions();
```

Comment:
Remove all use of `RuntimeHelpers.PrepareConstrainedRegions();` and `Thread.Begin/EndCriticalRegion();`

Reply:
Removed